stdio/file.c

```diff
@@ -855,9 +872,25 @@ void rewind(FILE *file)
 }
 
 
+int fseeko_unlocked(FILE *stream, off_t offset, int whence)
+{
+	return (int)fseektell_unlocked(stream, offset, whence);
+}
+
+
 int fseeko(FILE *stream, off_t offset, int whence)
 {
-	return fseek(stream, offset, whence);
+	int ret;
+	mutexLock(stream->lock);
+	ret = fseeko_unlocked(stream, offset, whence);
+	mutexUnlock(stream->lock);
+	return (ret >= 0) ? 0 : -1;
+}
+
+
+int fsetpos(FILE *stream, const fpos_t *pos)
+{
+	return (fseeko(stream, *(const off_t *)pos, SEEK_SET) >= 0) ? 0 : -1;
```

Comment:
This cast would become unsafe should `sizeof(off_t) != sizeof(fpos_t)`.
```suggestion
	off_t off = (off_t)(*pos);
	return (fseeko(stream, off, SEEK_SET) >= 0) ? 0 : -1;
```

Reply:
I agree with you that an intermediate variable on the stack would be safer. But what appeals to me more is why the direct cast might be unreadable: "On  some non-NIX systems, an `fpos_t` object may have a complex representation", and on my host it's actually a structure actually.

I've checked that, in case of `libphoenix` currently is true:
`sizeof(off_t) == sizeof(fpos_t)`

##

`off_t` is `long long int` a.k.a. `off64_t`
https://github.com/phoenix-rtos/libphoenix/blob/b2bbdb12dccc098f32981f761ada8d6e05af7eb0/include/sys/types.h#L41-L42

##

`fpos_t` is `long long int` a.k.a. `offs_t` (on all arch)
https://github.com/phoenix-rtos/libphoenix/blob/b2bbdb12dccc098f32981f761ada8d6e05af7eb0/include/stdio.h#L61

https://github.com/phoenix-rtos/libphoenix/blob/b2bbdb12dccc098f32981f761ada8d6e05af7eb0/include/arch/ia32/types.h#L20
https://github.com/phoenix-rtos/libphoenix/blob/b2bbdb12dccc098f32981f761ada8d6e05af7eb0/include/arch/armv7/types.h#L20
https://github.com/phoenix-rtos/libphoenix/blob/b2bbdb12dccc098f32981f761ada8d6e05af7eb0/include/arch/arm-imx/types.h#L20
https://github.com/phoenix-rtos/libphoenix/blob/b2bbdb12dccc098f32981f761ada8d6e05af7eb0/include/arch/riscv64/types.h#L20

Reply:
If `fpos_t` would become a structure, casting a pointer will cause a nasty bug (hopefully a warning too)